### DISCUSSION TITLE
CharacterVirtual - porting of CharacterVirtualTest logic

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#23] [Show and tell]
It required some additional (and marking optional) IDL methods to match the logic, and I had to map some of the methods to ThreeJS Vector3/Quaternion, but I got an initial stab at the CharacterVirtualTest working.

I'm still working out how to adjust the Three Models to match exactly the collision with the characters, and I will need to start stubbing out some hybrid JSImplementation if I want to have the collision callbacks in JS to match some of the sliding and conveyer-belt logic.
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/WebIDL-Binder.html#sub-classing-c-base-classes-in-javascript-jsimplementation

Fork: https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/character-virtual-example

So far I have been testing stair height and ramp incline thresholds and have a poor keyboard interface setup for moving and jumping.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/fdd899ad-2fc7-40a0-bddc-db2b11a48853)

**Comment by jrouwe** [comment]
That's looking good! Thanks for doing this! Send me a PR if you're happy with it.

Collision callbacks are indeed an area that I still needed to investigate, but even without that it looks like a good demo.

**Comment by PhoenixIllusion** [comment]
Updated currently to have camera that follows the character. Fixed the player sizes to match exactly the JoltPhysics collision.

I setup a CharacterContactListenerJS in the IDL and was testing it.
I got conveyor belt to work (blue box). I wired a large box under the level as teleport back to the level using OnContactValidate (not sure if the correct callback).

I do not yet have OnContactAdded or OnContactSolve triggering for some reason. Once I get those, I may add one a pushing object or pushable object to finalize my test level.

![2023-10-04-jolt-js-char-test](https://github.com/jrouwe/JoltPhysics.js/assets/756488/45c68be4-8c69-457c-80c2-dfc5ed0b69ae)

**Reply by PhoenixIllusion** [reply]
Quite a pain on the OnContactAdded and Solve. If you get the WebIDL signature wrong (is a Vec3 by pointer or Ref) it will fail the signature check and assume you are making an alternate method with a diff signature, and instead call the method of the inherited base class listener, which is a NOP.

**Reply by PhoenixIllusion** [reply]
I can't get the signatures to match. The passed in Vec3's are not [Vec3 *] or [Vec3 &], they are objects. The WebIDL doesn't allow passing in objects of just plain [Vec3]

**Comment by PhoenixIllusion** [comment]
Wrote an intermediate wrapper on the Character Listener that just bounces to a WebIDL compatible version of the 2 methods. Got it working and added in a test of the pushable block. I don't have Kinetic blocks in my sample yet, but may be ready for PR.

![2023-10-04-physics-js-push-block](https://github.com/jrouwe/JoltPhysics.js/assets/756488/f9aa3586-f863-40cb-9520-2f8686c78bb7)
